util/topgen/intermodule.py

```diff
@@ -0,0 +1,111 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+from typing import Dict
+
+from .lib import *
+
+
+def elab_intermodule(topcfg):
+    """Check the connection of inter-module and categorize them
+
+    In the top template, it uses updated inter_module fields to create
+    connections between the modules (incl. memories). This function is to
+    create and check the validity of the connections `inter_module` using IPs'
+    `inter_signal_list`.
+    """
+
+    list_of_intersignals = []
+
+    if "inter_signal" not in topcfg:
+        topcfg["inter_signal"] = {}
+
+    # Gather the inter_signal_list
+    instances = topcfg["module"] + topcfg["memory"]
+
+    intermodule_instances = [x for x in instances if "inter_signal_list" in x]
+
+    for x in intermodule_instances:
+        for sig in x["inter_signal_list"]:
+            # Add instance name to the entry and add to list_of_intersignals
+            sig["inst_name"] = x["name"]
+            list_of_intersignals.append(sig)
+
+    # TODO: Cross check
+
+    # Add field to the topcfg
+    topcfg["inter_signal"]["signals"] = list_of_intersignals
+
+    # intermodule
+    definitions = []
+
+    uid = 0  # Unique connection ID across the top
+
+    for req, rsps in topcfg["inter_module"].items():
+        log.info("{req} --> {rsps}".format(req=req, rsps=rsps))
+        req_module, req_signal = req.split('.')
+
+        # get the module signal
+        req_struct = find_intermodule_signal(list_of_intersignals, req_module,
+                                             req_signal)
+
+        rsp_len = len(rsps)
+        for i, rsp in enumerate(rsps):
+            assert i == 0, "Handling multiple connections (array) isn't yet supported"
```

Comment:
this is a nit, but wouldn't it be better for now to just assert rsp_len is always 1 instead of checking it whether the loop has possibly advanced?

Reply:
I think regardless of it, as the definition of inter-module connection is from request to multiple responses,the for loop should be here...

Wait.. I might be wrong about the definition , I considered there's array of requests which spread out to responses. But it might be a case that multiple requests go into an array of requests (single port).

I will think about it more.. maybe the structure should be changed...

Reply:
o i'm sorry i think i misspoke. I just meant the temporary check because we only currently handle 1:1 can be outside as a check on rsp_len, instead of checking the loop index.

Maybe i misunderstood something else?